### PR TITLE
[WIP] Test Implementation of OpenTime in C++

### PR DESCRIPTION
**Pull request by ssteinbach**
Hi all, this pull request adds a new subdirectory, `cpp_api` in which a test implementation of opentime in C++ resides, along with a pybind11 python binding.  I've also copied in the `test_opentime.py` suite so you can see what changes had to be made in order to make things pass.  This is intended more to draw out conversation and feedback, and to serve as a reference point for conversations with folks we're having about a C++ API.  Let us know if you have any thoughts!

Before this can be merged in:
- need to integrate or replace the CMake build system with the top level setup.py
- to to move this library so that the resulting library gets put in a place where the normal `opentimelineio` module can find it
- There are some outstanding TODOs
- Should make decisions about memory management in the library and other low level policy decisions that C++ draws out (extern c bindings, exceptions, etc).

Thanks everyone!

**Quick Reminder** If you're curious about some of the previous discussion around this issue, see:
https://github.com/PixarAnimationStudios/OpenTimelineIO/wiki/Wrapping-OTIO-C-or-C-Plus-Plus-In-Python

### CODE REVIEW

File: cpp_api/opentime.h [hunk 1]
```diff
@@ -0,0 +1,910 @@
...
+    double time_obj_value = time_obj.value;
+    double time_obj_rate = time_obj.rate;
+
+    rt_value_t total_seconds = to_seconds(time_obj);
+
+    // @TODO: fun fact, this will print the wrong values for numbers at a certain 
+    // number of decimal places, if you just std::cerr << total_seconds
+    /* DEBUG_PRINT(total_seconds); */
+
+    auto try2 = time_obj.value_rescaled_to(1.0);
```
Comment (jminor) [inline cpp_api/opentime.h]: try2?

File: cpp_api/opentime.cpp [hunk 2]
```diff
@@ -0,0 +1,656 @@
...
+    if (not rate_is_dropframe)
+    {
+        if (std::round(rate) == 24)
+        {
+            rate = 24.0;
+        }
+    }
+    else
+    {
+        if (rate == static_cast<rt_rate_t>(29.97))
```
[the remark this answers is not in the document]
Comment (ssteinbach) [reply cpp_api/opentime.cpp]: Cool, good call. At the moment, this is what the python implementation is doing, so if its wrong here, its wrong there too.  Do you have any suggestions about what that more subtle thing might be?

File: cpp_api/opentime.cpp [hunk 3]
```diff
@@ -0,0 +1,656 @@
...
+namespace std 
+{
+    size_t 
+    hash<opentime::RationalTime>::operator()(
+            const opentime::RationalTime& rt
+    ) const 
+    {
+            return (
+                    (std::hash<opentime::rt_value_t>{}(rt.value)) 
+                    ^ ((std::hash<opentime::rt_rate_t>{}(rt.rate)) >> 1)
```
[the remark this answers is not in the document]
Comment (ssteinbach) [reply cpp_api/opentime.cpp]: Awesome, I'll snag that.  There is another question: do you think these types should be hashable?  I'm open to making them not hashable.

File: cpp_api/opentime.h [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+/// type wrappers for rational time
+using rt_value_t=double;
+using rt_rate_t=double;
+
+/// ensure that the rate is in one of the valid timecode rate lists
+bool validate_timecode_rate(const rt_rate_t rate);
+
+/// A point in time, value*rate samples after 0.
+class RationalTime
```
[the remark this answers is not in the document]
Comment (ssteinbach) [reply cpp_api/opentime.h]: Yeah, we've talked about that but haven't built that out yet.  I'm open to that as well.  Thanks!

**Comment by ssteinbach** [comment]
See #391

**Pull request closed without merging**